### PR TITLE
Expose http sources for metrics in values.yaml

### PR DESCRIPTION
**Pull request by sumo-drosiek**
###### Description

Expose http sources in values.yaml

###### ToDO

 - [x] Move out configuration from configmap to separate files (like fluentd conf)
 - [x] Discuss if logs/events/default metrics should be exposed
        I decided to not expose logs and events so far
 - [x] Add enable/disable flag
 - [x] Dynamic generation of fluentd configuration(?)

###### Testing performed

- [ ] ci/build.sh
- [ ] Redeploy fluentd and fluentd-events pods
- [ ] Confirm events, logs, and metrics are coming in

### CODE REVIEW

File: deploy/helm/sumologic/templates/setup/setup-configmap.yaml [hunk 1]
```diff
@@ -81,52 +69,18 @@ data:
...
+    {{ range $source := .Values.sumologic.sources }}
+    resource "sumologic_http_source" "{{ template "sources.terraform_name" $source }}" {
+        name         = local.{{ template "sources.terraform_source_name" $source }}
         collector_id = "${sumologic_collector.collector.id}"
     }
+    {{ end -}}
 
     resource "sumologic_http_source" "events_source" {
         name         = local.events-source-name
         category     = {{ if .Values.fluentd.events.sourceCategory }}{{ .Values.fluentd.events.sourceCategory | quote }}{{- else}}{{ "\"${var.cluster_name}/${local.events-source-name}\"" }}{{- end}}
```
Comment (sumo-drosiek) [inline deploy/helm/sumologic/templates/setup/setup-configmap.yaml:81]: move category inside values.yaml

File: deploy/kubernetes/setup-sumologic.yaml.tmpl [hunk 2]
```diff
@@ -146,15 +155,16 @@ data:
...
-        endpoint-events                           = "${sumologic_http_source.events_source.url}"
-        endpoint-logs                             = "${sumologic_http_source.logs_source.url}"
         endpoint-metrics                          = "${sumologic_http_source.default_metrics_source.url}"
         endpoint-metrics-apiserver                = "${sumologic_http_source.apiserver_metrics_source.url}"
         endpoint-metrics-kube-controller-manager  = "${sumologic_http_source.kube_controller_manager_metrics_source.url}"
         endpoint-metrics-kube-scheduler           = "${sumologic_http_source.kube_scheduler_metrics_source.url}"
         endpoint-metrics-kube-state               = "${sumologic_http_source.kube_state_metrics_source.url}"
         endpoint-metrics-kubelet                  = "${sumologic_http_source.kubelet_metrics_source.url}"
         endpoint-metrics-node-exporter            = "${sumologic_http_source.node_exporter_metrics_source.url}"
+        endpoint-control-plane                    = "${sumologic_http_source.control_plane_metrics_source.url}"
```
Comment (perk-sumo) [inline deploy/kubernetes/setup-sumologic.yaml.tmpl:165]: It's back, good :)

File: deploy/helm/sumologic/values.yaml [hunk 3]
```diff
@@ -390,6 +420,58 @@ fluentd:
...
+      control-plane:
+        match: prometheus.metrics.control-plane**
+        id: sumologic.endpoint.metrics.control.plane
+        endpoint: metrics-control-plane
+        weight: 90
+      default:
+        match: prometheus.metrics**
+        id: sumologic.endpoint.metrics
+        endpoint: metrics
+        weight: 100
```
Comment (perk-sumo) [inline deploy/helm/sumologic/values.yaml:473]: I like this 👍

File: deploy/helm/sumologic/values.yaml [hunk 4]
```diff
@@ -82,6 +82,36 @@ sumologic:
         helm.sh/hook-weight: "0"
         helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
 
+  ## Configuration of http sources
+  ## name: source name
+  ## value: source values - it's visible in sumologic platform
+  ## endpoint: binds source with fluentd configuration: (fluentd.metrics.output[])
+  sources:
```
Comment (perk-sumo) [inline deploy/helm/sumologic/values.yaml:89]: This would better be a default sources + additional sources. Or something like that.

**Comment by sumo-drosiek** [comment]
I updated the PR with more user-friendly configuration, but this is highly breaking change for non-helm users

**Comment by sumo-drosiek** [comment]
Here is example of adding new source to sumologic and fluentd:

```
    --set sumologic.sources.new_source.value="my new source" \ # create new source named my new source and available in helm as new_source
    --set fluentd.metrics.output.new_source.match="test_match" \ # fluentd <match test_match> clausule
    --set fluentd.metrics.output.new_source.id="test_id" \ # fluentd @id parameter
    --set fluentd.metrics.output.new_source.weight=0 \ # order of fliuentd match (lower, earlier)
    --set fluentd.metrics.output.new_source.drop=true # if true, drop traffic instead of passing it to the sumologic \
    --set fluentd.buffer.filePaths.metrics.new_source=/fluentd/buffer/metrics.apiserver # set persistence
```

**Comment by sumo-drosiek** [comment]
Unfortunately if source value is changed, the new source is created via terraform and old one is not cleaned up

File: deploy/helm/prometheus-overrides.yaml [hunk 5]
```diff
@@ -228,13 +228,13 @@ prometheus:
...
       # coredns
-      url: http://$(CHART).$(NAMESPACE).svc.cluster.local:9888/prometheus.metrics.control-plane
+      url: http://$(CHART).$(NAMESPACE).svc.cluster.local:9888/prometheus.metrics.control-plane.coredns
       writeRelabelConfigs:
       - action: keep
         regex: coredns
         sourceLabels: [job]
     - # etcd server
-      url: http://$(CHART).$(NAMESPACE).svc.cluster.local:9888/prometheus.metrics.control-plane
+      url: http://$(CHART).$(NAMESPACE).svc.cluster.local:9888/prometheus.metrics.control-plane.kube-etcd
```
Comment (sumo-drosiek) [inline deploy/helm/prometheus-overrides.yaml:237]: Fixes prometheus issue. It has been saying that configuration for `http://$(CHART).$(NAMESPACE).svc.cluster.local:9888/prometheus.metrics.control-plane` is duplicated

**Comment by sumo-drosiek** [comment]
@frankreno Do we consider fluentd env changes as breaking changes?

File: deploy/helm/prometheus-overrides.yaml (hunk 5 above)
Comment (frankreno) [reply deploy/helm/prometheus-overrides.yaml:237]: The only thing i find misleading with this setup is that we collect from other control plane components and do not prefix them with control-plane.  Any reason to just remove the control-plane part?

File: deploy/helm/sumologic/conf/events/events.conf [hunk 6]
```diff
@@ -35,7 +35,7 @@
   @type sumologic
   @id sumologic.endpoint.events
   sumo_client {{ include "sumologic.sumo_client" . | quote }}
-  endpoint "#{ENV['SUMO_ENDPOINT_EVENTS']}"
+  endpoint "#{ENV['SUMO_ENDPOINT_EVENTS_SOURCE']}"
```
Comment (frankreno) [inline deploy/helm/sumologic/conf/events/events.conf:38]: Would this be considered a breaking change for non-helm users? Helm should handle this, but what abotu non-helm.

File: deploy/helm/sumologic/values.yaml [hunk 7]
```diff
@@ -390,6 +412,52 @@ fluentd:
     ## To use additional output plugins
     extraOutputPluginConf: |-
 
+    ## Output configuration
+    ## match: fluentd match
```
Comment (frankreno) [inline deploy/helm/sumologic/values.yaml:416]: nit - we should call this tag imo because that is what it really is, the tag for the match block.

File: deploy/kubernetes/setup-sumologic.yaml.tmpl [hunk 8]
```diff
@@ -79,51 +83,56 @@ data:
...
         collector_id = "${sumologic_collector.collector.id}"
     }
-  
-    resource "sumologic_http_source" "node_exporter_metrics_source" {
-        name         = local.node-exporter-metrics-source-name
+    
+    resource "sumologic_http_source" "events_source" {
+        name         = local.events_source
         collector_id = "${sumologic_collector.collector.id}"
+        category     = "${var.cluster_name}/${local.events_source}"
```
Comment (frankreno) [inline deploy/kubernetes/setup-sumologic.yaml.tmpl:135]: this seems to revert the current behavior of being able to change the sourceCategory for events?

https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/master/deploy/helm/sumologic/conf/setup/sumo-k8s.tf#L56

File: deploy/kubernetes/setup-sumologic.yaml.tmpl [hunk 9]
```diff
@@ -79,51 +83,56 @@ data:
...
-    resource "sumologic_http_source" "kubelet_metrics_source" {
-        name         = local.kubelet-metrics-source-name
+    
+    resource "sumologic_http_source" "scheduler_metrics_source" {
+        name         = local.scheduler_metrics_source
         collector_id = "${sumologic_collector.collector.id}"
     }
-  
+    
+    resource "sumologic_http_source" "state_metrics_source" {
```
Comment (frankreno) [inline deploy/kubernetes/setup-sumologic.yaml.tmpl:122]: Perhaps we do this in another PR, but I believe we need to expose the sourceCategory for all metrics sources.  Right now, we do not set it which results in all metrics getting the default sourceCageory of "Http Input". this makes data volume reporting tricky.  We should follow the same model as events for the metrics sources so that the data volume reporting in Sumo is managemble.

Comment (frankreno) [reply deploy/kubernetes/setup-sumologic.yaml.tmpl:122]: Logs are a bit different, because logs sourceCategory is set via the HTTP header. This is not true for metrics and events which is why we need to be able to provide a sensible default.

File: deploy/kubernetes/setup-sumologic.yaml.tmpl (hunk 8 above)
Comment (samjsong) [reply deploy/kubernetes/setup-sumologic.yaml.tmpl:135]: This is actually consistent with our current behavior since this file is for non-helm users

https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/master/deploy/kubernetes/setup-sumologic.yaml.tmpl#L95

File: deploy/helm/sumologic/conf/events/events.conf (hunk 6 above)
Comment (samjsong) [reply deploy/helm/sumologic/conf/events/events.conf:38]: actually since non-helm users just run `kubectl apply -f fluentd-sumologic.yaml` to upgrade, the new yaml file should have the renamed environment variables so it should be fine.

However there IS a breaking change here, and that is a result of the keys in the sumologic secret being changed (added suffix `-source`), which means that non-helm users will have to re-run the setup job in order to get an updated secret (though the collector and sources will remain untouched).

**Comment by samjsong** [comment]
Overall LGTM, one thing that might be cool to add (probably in separate PR) is the ability to append more sources. In fact the Zaidan team has the need for additional http sources and in my migration work I don't think I will be able to move everything until we have this functionality

File: deploy/helm/prometheus-overrides.yaml (hunk 5 above)
Comment (sumo-drosiek) [reply deploy/helm/prometheus-overrides.yaml:237]: Yes, the fluentd configuration will be more complex: either more complex regex for tags or output section per endpoint

**Comment by sumo-drosiek** [comment]
@samjsong What do you mean by additional http sources?

I think the `sumologic.sources` section should meet you expectations :)

File: deploy/helm/prometheus-overrides.yaml (hunk 5 above)
Comment (frankreno) [reply deploy/helm/prometheus-overrides.yaml:410]: How many metrics is each of these producing? We need to be conscious of blacklisting and if we are close to the limit, we should distribute across multiiple HTTP sources to prevent that.

File: deploy/helm/sumologic/conf/events/events.conf (hunk 6 above)
Comment (frankreno) [reply deploy/helm/sumologic/conf/events/events.conf:38]: Good catch, yeah we dont what the breaking change from the secret.

File: deploy/helm/sumologic/values.yaml (hunk 7 above)
Comment (sumo-drosiek) [reply deploy/helm/sumologic/values.yaml]: fixed

**Comment by sumo-drosiek** [comment]
@frankreno We are planning to reduce the metrics list anyway. Waiting for dashboards and exact list of metrics

**Comment by vsinghal13** [comment]
Also, one more suggestion, it is best practice to have separate `tf` files for variables, resources, and providers. Would be great if we can split `sumo-k8s.tf` similarly.

File: deploy/helm/sumologic/templates/_helpers.tpl [hunk 10]
```diff
@@ -353,27 +353,157 @@ Generate metrics match configuration
...
+{{- if .Drop }}
+  @type null
+{{- else }}
   @type sumologic
   @id {{ .Id }}
-  endpoint "#{ENV['{{ .Endpoint }}']}"
-{{- .Values.fluentd.metrics.outputConf | nindent 2 }}
+  sumo_client {{ include "sumologic.sumo_client" .Context | quote }}
+  endpoint "#{ENV['{{ include "terraform.sources.endpoint" .Endpoint}}']}"
+{{- .Context.Values.fluentd.metrics.outputConf | nindent 2 }}
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply deploy/helm/sumologic/templates/_helpers.tpl:372]: I need to pass k8s context to have access to .Values, `.Context` in this template is simply `.` from kubernetes. 
`.` is override in `range` functions and templates, so there is need to pass it as a variable (or a dict value for templates, because they get only one argument). Here is flow (btw it covers both cases):
https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/8e58e3f3f1da3a704e7b2a70bd166c1cd9aa1012/deploy/helm/sumologic/conf/metrics/metrics.conf#L50
https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/8e58e3f3f1da3a704e7b2a70bd166c1cd9aa1012/deploy/helm/sumologic/conf/metrics/metrics.conf#L63

File: deploy/kubernetes/fluentd-sumologic.yaml.tmpl [hunk 11]
```diff
@@ -389,107 +389,114 @@ data:
...
-    
-      <match prometheus.metrics.kubelet**>
+      
+      <match prometheus.metrics.container**>
         @type sumologic
-        @id sumologic.endpoint.metrics.kubelet
-        endpoint "#{ENV['SUMO_ENDPOINT_METRICS_KUBELET']}"
+        @id sumologic.endpoint.metrics.container
+        sumo_client "k8s_1.0.0"
+        endpoint "#{ENV['SUMO_ENDPOINT_KUBELET_METRICS_SOURCE']}"
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply deploy/kubernetes/fluentd-sumologic.yaml.tmpl:408]: I suppose we have an exception here. For 1.0 release we are using kubelet endpoint for container.
https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/bd42eaeaa7da7b5d2dd64c9e81a5bef69214c097/deploy/helm/sumologic/conf/metrics/metrics.conf#L73-L88

**Comment by sumo-drosiek** [comment]
@vsinghal13 I will address split of the tf file in the another PR. This PR is oversized already 😞

File: deploy/helm/sumologic/values.yaml [hunk 12]
```diff
@@ -429,6 +458,52 @@ fluentd:
     #     mountPath: /certs
     #     readOnly: true
   
+    ## Output configuration
+    ## tag: tag for fluentd match
+    ## id: sumologic output @id
+    ## endpoint: [optional] [output key by default] It should match the sumologic.sources[].endpoint
+    ## drop: [optional] [false by default] Change it to true to drop traffic for specific output
+    ## weight: [optional [0 by default] Order of fluentd output (lower means higher priority)
+    ## The weight has meaning in case of fluentd matches
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply deploy/helm/sumologic/values.yaml:467]: ok, I will address it in documentation rather than comment in the values.yaml, but here we go:

Let's say we have two two sources (`source1`, `source2`)

```
      source1:
        tag: prometheus.metrics.custom**
        weight: 90
      source2:
        tag: **
        weight: 80
```
is going to generate in fluentd 
```
# weight 80
<match **>
  # takes all records
</match>
# weight 90
<match prometheus.metrics.custom**>
  # it won't work
</match>
```

but if we switch the weights:

```
      source1:
        tag: **
        weight: 80
      source2:
        tag: prometheus.metrics**
        weight: 90
```
then flunetd config will change to
```
# weight 80
<match prometheus.metrics.custom**>
  # takes custom records
</match>
# weight 90
<match **>
  # takes all other records
</match>
```

[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply deploy/helm/sumologic/values.yaml:467]: I'm going to address it in https://github.com/SumoLogic/sumologic-kubernetes-collection/issues/680#issuecomment-634607171

[the remark this answers is not in the document]
Comment (vsinghal13) [reply deploy/helm/sumologic/values.yaml:466]:
```# weight 80
<match **>
  # takes all records
</match>
# weight 90
<match prometheus.metrics.custom**>
  # it won't work
</match>```

Even if the user specified this, the second match won't work because everything was already matched by the catch all block, right?
